### PR TITLE
Add IOperation support for C# patterns

### LINKED ISSUE #19927: IOperation support for patterns
Proposed API:

```
    public interface IPattern : IOperation
    {
    }

    public interface IDeclarationPattern : IPattern
    {
        ISymbol DeclaredSymbol { get; }
    }

    public interface IConstantPattern : IPattern
    {
        IOperation Value { get; }
    }

    public interface IPatternCaseClause : ICaseClause
    {
        ILabelSymbol Label { get; }
        IPattern Pattern { get; }
        IOperation GuardExpression { get; }
    }

    public interface IIsPatternExpression : IOperation
    {
        IOperation Expression { get; }
        IPattern Pattern { get; }
    }
```

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #19927

### CODE REVIEW

**Comment by mavasani** [comment]
Tagging @dotnet/analyzer-ioperation @dotnet/roslyn-compiler for review.

File: src/Compilers/Core/Portable/Operations/OperationKind.cs [hunk 1]
```diff
@@ -144,6 +144,8 @@ public enum OperationKind
         // 0x205 is open for future use.
         /// <summary>Indicates an <see cref="IIncrementExpression"/>.</summary>
         IncrementExpression = 0x206,
+        /// <summary>Indicates an <see cref="IIsPatternExpression"/>.</summary>
+        IsPatternExpression = 0x207,
```
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/OperationKind.cs]: We should fill 0x205

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 2]
```diff
@@ -179,6 +179,18 @@ private IOperation CreateInternal(BoundNode boundNode)
                     return CreateBoundInterpolationOperation((BoundStringInsert)boundNode);
                 case BoundKind.LocalFunctionStatement:
                     return CreateBoundLocalFunctionStatementOperation((BoundLocalFunctionStatement)boundNode);
+                case BoundKind.ConstantPattern:
+                    return CreateBoundConstantPatternOperation((BoundConstantPattern)boundNode);
+                case BoundKind.DeclarationPattern:
+                    return CreateBoundDeclarationPatternOperation((BoundDeclarationPattern)boundNode);
+                case BoundKind.WildcardPattern:
+                    return null;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: What's up with this?

Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: This bound node is created for a default label within pattern switch statement. The bound pattern case clause which is its parent already exposes the label symbol for the default label. In the design meeting we decided that we will represent such a default label within pattern switch statement as a null pattern. See unit tests added for default label in pattern case clause for tree structure.

File: src/Compilers/Core/Portable/Operations/OperationKind.cs (hunk 1 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/Operations/OperationKind.cs]: Sure, I can adjust it...

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+                // CS0266: Cannot implicitly convert type 'double' to 'int?'. An explicit conversion exists (are you missing a cast?)
+                //         if (/*<bind>*/x is 12.0/*</bind>*/) Console.WriteLine(y);        
+                Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "12.0").WithArguments("double", "int?").WithLocation(8, 28)
+            };
+
+            VerifyOperationTreeAndDiagnosticsForTest<IsPatternExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [Fact, WorkItem(19927, "https://github.com/dotnet/roslyn/issues/19927")]
+        public void TestIsPatternExpression_ConstantPatternWithNoConversion()
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs]: There is an implicit conversion here. Can we add a test for something that really shouldn't have a conversion, like just 12, and rename this to reflect the implicit conversion?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+                // CS0246: The type or namespace name 'UndefinedType' could not be found (are you missing a using directive or an assembly reference?)
+                //         if (/*<bind>*/x is UndefinedType y/*</bind>*/) Console.WriteLine(y);        
+                Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "UndefinedType").WithArguments("UndefinedType").WithLocation(8, 28)
+            };
+
+            VerifyOperationTreeAndDiagnosticsForTest<IsPatternExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [Fact, WorkItem(19927, "https://github.com/dotnet/roslyn/issues/19927")]
+        public void TestIsPatternExpression_NonConstantExpressionInsteadOfPatternDeclaration()
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs]: I'm not sure what this test name is tell me? Especially since there is an IConstantPattern in the resulting tree.

Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs]: I will rename it appropriately, the case label needs to be a constant label or a declaration pattern, it cannot be a constant pattern in valid code.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs (hunk 3 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs]: Yes, sure.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs [hunk 5]
```diff
@@ -0,0 +1,679 @@
...
+                break;
+        }
+    }
+}
+";
+            string expectedOperationTree = @"
+ISwitchStatement (1 cases) (OperationKind.SwitchStatement) (Syntax: 'switch (x) ... }')
+  Switch expression: ILocalReferenceExpression: x (OperationKind.LocalReferenceExpression, Type: System.Int32?) (Syntax: 'x')
+  Sections: ISwitchCase (1 case clauses, 1 statements) (OperationKind.SwitchCase) (Syntax: 'case var y: ... break;')
+        Clauses: IPatternCaseClause (Label Symbol: case var y:) (OperationKind.PatternCaseClause) (Syntax: 'case var y:')
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: Not part of this PR, but can we file a but to make the clauses be on new lines and indented? Makes it harder to tell that the Pattern is actually a child of the IPatternCaseClause, not the ISwitchCase.

Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: We already have an issue for this and @jinujoseph has asked the vendors to fix this as soon as possible. Compiler team has requested the same multiple times to ease the reviews.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs [hunk 6]
```diff
@@ -0,0 +1,679 @@
...
+            case null:
+                break;
+            /*<bind>*/case X y:/*</bind>*/
+                break;
+        }
+    }
+}
+";
+            string expectedOperationTree = @"
+ISwitchStatement (2 cases) (OperationKind.SwitchStatement) (Syntax: 'switch (x) ... }')
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: Will this test be relevant once #18042 is fixed?

Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: Yes, we still want to validate the IOperation tree structure for the pattern case label when the switch statement also has a constant expression label. We will need to edit the baselines for almost all tests in this file though.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs [hunk 7]
```diff
@@ -0,0 +1,679 @@
...
+}
+";
+            string expectedOperationTree = @"
+ISwitchStatement (1 cases) (OperationKind.SwitchStatement, IsInvalid) (Syntax: 'switch (x) ... }')
+  Switch expression: IParameterReferenceExpression: x (OperationKind.ParameterReferenceExpression, Type: System.Int32?) (Syntax: 'x')
+  Sections: ISwitchCase (1 case clauses, 2 statements) (OperationKind.SwitchCase, IsInvalid) (Syntax: 'case /*</bi ... break;')
+        Clauses: ISingleValueCaseClause (Equality operator kind: BinaryOperationKind.IntegerEquals) (CaseKind.SingleValue) (OperationKind.SingleValueCaseClause, IsInvalid) (Syntax: 'case /*</bind>*/')
+            IConversionExpression (ConversionKind.Invalid, Implicit) (OperationKind.ConversionExpression, Type: System.Int32, IsInvalid) (Syntax: '')
+              IInvalidExpression (OperationKind.InvalidExpression, Type: ?, IsInvalid) (Syntax: '')
+        Body: IVariableDeclarationStatement (1 declarations) (OperationKind.VariableDeclarationStatement) (Syntax: 'const int y')
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: Is this actually what we're expecting here? It certainly wasn't what I was expecting to see.

Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: Its broken code, and compiler ended up parsing `const int y` as a variable declaration statement inside the switch section body.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >return null; [](start = 20, length = 12)

Do we have a test that hits this code path?

Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Yes, see this test: https://github.com/dotnet/roslyn/pull/20276/files#diff-7d83c86cd85172860c4416b70dc793eaR450. The default label here is bound as a BoundWildCardPattern.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs [hunk 8]
```diff
@@ -0,0 +1,679 @@
...
+    }
+}
+";
+            string expectedOperationTree = @"
+ISwitchStatement (1 cases) (OperationKind.SwitchStatement) (Syntax: 'switch (x) ... }')
+  Switch expression: IParameterReferenceExpression: x (OperationKind.ParameterReferenceExpression, Type: System.Object) (Syntax: 'x')
+  Sections: ISwitchCase (2 case clauses, 1 statements) (OperationKind.SwitchCase) (Syntax: 'case X y: ... break;')
+        Clauses: IPatternCaseClause (Label Symbol: case X y:) (OperationKind.PatternCaseClause) (Syntax: 'case X y:')
+            Pattern: IDeclarationPattern (Declared Symbol: X y) (OperationKind.DeclarationPattern) (Syntax: 'X y')
+          IPatternCaseClause (Label Symbol: default:) (OperationKind.PatternCaseClause) (Syntax: 'default:')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: >IPatternCaseClause (Label Symbol: default:) (OperationKind.PatternCaseClause) (Syntax: 'default:') [](start = 10, length = 98)

I believe we decided to use regular default case clause rather than a IPatternCaseClause with nulls.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: >IPatternCaseClause (Label Symbol: default:) (OperationKind.PatternCaseClause) (Syntax: 'default:') [](start = 10, length = 98)

The test output should include all members of the node, even those with null values. Reviewers and readers shouldn't guess what pieces of information are there with null values.

Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: I can switch the IOperation tree to use this approach. Do you recommend going a step further and change the binder to not even have a `BoundKind.WildcardPattern` if we are just treating it as a regular default label everywhere?

Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: @jinujoseph This is another recommendation to enhance the IOperation test output dump. Currently we skip dump for null child nodes.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: > This is another recommendation to enhance the IOperation test output dump. 

Ppease fix it for nodes added in this PR as part of this PR.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: > Do you recommend going a step further and change the binder to not even have a BoundKind.WildcardPattern

No, I do not recommend that, the labels in BoundPatternSwitchSection are strongly typed. After looking at this issue more, it looks like we don't have a dedicated IOperation node to represent a regular ```default``` label. It is represented as a label with null value. Is this correct? In this case, I think the current representation is fine, I was hoping to avoid null properties.

Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: Sure, I can do so.

Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: We do not have a dedicated IOperation for default label, but do have a different case kind: http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/CaseKind.cs,65f316f2e0138ff5,references. Current approach now seems wrong to me as the case kind will be Pattern for the default label, it should instead be Default, which will be the case with your original suggestion (I also need to update the dump to also print the case kind).

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.cs [hunk 9]
```diff
@@ -657,16 +657,15 @@ public void M(int x)
...
-            IBinaryOperatorExpression (BinaryOperationKind.IntegerGreaterThanOrEqual) (OperationKind.BinaryOperatorExpression, Type: System.Boolean) (Syntax: 'x >= 10')
-              Left: IParameterReferenceExpression: x (OperationKind.ParameterReferenceExpression, Type: System.Int32) (Syntax: 'x')
-              Right: ILiteralExpression (Text: 10) (OperationKind.LiteralExpression, Type: System.Int32, Constant: 10) (Syntax: '10')
-        IBranchStatement (BranchKind.Break) (OperationKind.BranchStatement) (Syntax: 'break;')
+ISwitchStatement (1 cases) (OperationKind.SwitchStatement) (Syntax: 'switch (x) ... }')
+  Switch expression: IParameterReferenceExpression: x (OperationKind.ParameterReferenceExpression, Type: System.Int32) (Syntax: 'x')
+  Sections: ISwitchCase (1 case clauses, 1 statements) (OperationKind.SwitchCase) (Syntax: 'case var y  ... break;')
+        Clauses: IPatternCaseClause (Label Symbol: case var y when (x >= 10):) (OperationKind.PatternCaseClause) (Syntax: 'case var y  ...  (x >= 10):')
+            Pattern: IDeclarationPattern (Declared Symbol: System.Int32 y) (OperationKind.DeclarationPattern) (Syntax: 'var y')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.cs]: >Declared Symbol: System.Int32 y [](start = 42, length = 31)

What node owns this local? It feels like switch should have a list of locals it owns and pattern section should have a list of locals it owns (similar to IBlockStatement).

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs (hunk 8 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: Alternatively, we can IPatternCaseClause have case kind Default for this scenario.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: Perhaps we should add a dedicated node to represent a ```default``` label, same node for both switch flavors?

Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: Let me open a separate issue to discuss all three possibilities in the design meeting. For now, I will take the approach 1 that you originally recommended.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: > Let me open a separate issue to discuss all three possibilities in the design meeting. For now, I will take the approach 1 that you originally recommended.

Sounds good.

Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: Filed https://github.com/dotnet/roslyn/issues/20312

**Comment by AlekseyTs** [comment]
CC @gafter

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.cs (hunk 9 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.cs]: The declared symbol here is owned by the IDeclarationPattern node. We can surely add extension methods or properties off IPatternCaseClause and ISwitchStatement to expose the locals in the corresponding scope. Let me file an issue to track the design and implementation here.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+    {
+        int? x = 12;
+        if (/*<bind>*/x is var y/*</bind>*/) Console.WriteLine(y);        
+    }
+}
+";
+            string expectedOperationTree = @"
+IIsPatternExpression (OperationKind.IsPatternExpression, Type: System.Boolean) (Syntax: 'x is var y')
+  Expression: ILocalReferenceExpression: x (OperationKind.LocalReferenceExpression, Type: System.Int32?) (Syntax: 'x')
+  Pattern: IDeclarationPattern (Declared Symbol: System.Int32? y) (OperationKind.DeclarationPattern) (Syntax: 'var y')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs]: >Declared Symbol: System.Int32? y [](start = 32, length = 32)

We should also test what node owns this local. I am suggesting to do this separately, rather than expanding the tested syntax here. We should test that for all IDeclarationPattern nodes.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.cs (hunk 9 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.cs]: > The declared symbol here is owned by the IDeclarationPattern node.

I do not believe this is consistent with handling of IDeclarationPattern nodes outside of a switch section. This node represents a point of declaration, it does not represent a scope for the local.

Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.cs]: Ah alright, I get your point. The DeclaredSymbol property here does not define scope ownership. Let me create a separate issue to discuss how to expose these locals in their owning scopes.

Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.cs]: Filed https://github.com/dotnet/roslyn/issues/20313

**Comment by AlekseyTs** [comment]
Many compiler tests for patterns verify SemanticModel behavior by using VerifyModelForDeclarationPattern test helper. Consider expanding that helper to perform validation for IOperation tree as well.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+{
+    void M(string x = /*<bind>*/string.Empty is string y/*</bind>*/)
+    {    
+    }
+}
+";
+            string expectedOperationTree = @"
+IIsPatternExpression (OperationKind.IsPatternExpression, Type: System.Boolean) (Syntax: 'string.Empty is string y')
+  Expression: IFieldReferenceExpression: System.String System.String.Empty (Static) (OperationKind.FieldReferenceExpression, Type: System.String) (Syntax: 'string.Empty')
+  Pattern: IDeclarationPattern (Declared Symbol: System.String y) (OperationKind.DeclarationPattern) (Syntax: 'string y')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs]: >Declared Symbol: System.String y [](start = 32, length = 32)

What node represents a scope for this declaration?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs [hunk 12]
```diff
@@ -0,0 +1,480 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    public partial class IOperationTests : SemanticModelTestBase
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs]: >IOperationTests [](start = 25, length = 15)

It would be good to add tests for field initializers, constructor initializers and attribute arguments.

**Comment by AlekseyTs** [comment]
There is also a VerifyModelForDeclarationField helper used by compiler tests to verify SemanticModel for declaration patters in script scenarios.

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs (hunk 10 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs]: This work is now tracked by #20313 - this tracks the design, implementation and testing for the locals exposed off owning nodes.

---
In reply to: [122779797](https://github.com/dotnet/roslyn/pull/20276#discussion_r122779797) [](ancestors = 122779797)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs (hunk 11 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs]: Work tracked with #20313

---
In reply to: [122784648](https://github.com/dotnet/roslyn/pull/20276#discussion_r122784648) [](ancestors = 122784648)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs (hunk 8 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: > Please fix it for nodes added in this PR as part of this PR.

The dumper is bailing out without dump in the common `OperationTreeVerifier.Visit(IOperation operation)` method, that affects all nodes. Lets track this work as part of the dumper improvement for all nodes.

---
In reply to: [122771355](https://github.com/dotnet/roslyn/pull/20276#discussion_r122771355) [](ancestors = 122771355)

**Comment by mavasani** [comment]
Discussed with Jinu and decided to add this test enhancement work to backlog, which the vendors can pick up later. Tracked by [https://github.com/dotnet/roslyn/issues/20332](https://github.com/dotnet/roslyn/issues/20332).

---
In reply to: [309521785](https://github.com/dotnet/roslyn/pull/20276#issuecomment-309521785) [](ancestors = 309521785)

**Comment by mavasani** [comment]
Discussed with Jinu and decided to add this test enhancement work to backlog, which the vendors can pick up later. Tracked by [https://github.com/dotnet/roslyn/issues/20332](https://github.com/dotnet/roslyn/issues/20332).

---
In reply to: [309525854](https://github.com/dotnet/roslyn/pull/20276#issuecomment-309525854) [](ancestors = 309525854)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs (hunk 12 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs]: Tests added at the end of the file.

---
In reply to: [122785158](https://github.com/dotnet/roslyn/pull/20276#discussion_r122785158) [](ancestors = 122785158)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: >return null; [](start = 20, length = 12)

Is this code path reachable now? Can we throw unreachable?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.cs [hunk 13]
```diff
@@ -693,20 +692,18 @@ public void M(int x)
...
+        Clauses: IPatternCaseClause (Label Symbol: case var y when (x >= 10):) (CaseKind.Pattern) (OperationKind.PatternCaseClause) (Syntax: 'case var y  ...  (x >= 10):')
+            Pattern: IDeclarationPattern (Declared Symbol: System.Int32 y) (OperationKind.DeclarationPattern) (Syntax: 'var y')
+            Guard Expression: IBinaryOperatorExpression (BinaryOperationKind.IntegerGreaterThanOrEqual) (OperationKind.BinaryOperatorExpression, Type: System.Boolean) (Syntax: 'x >= 10')
+                Left: IParameterReferenceExpression: x (OperationKind.ParameterReferenceExpression, Type: System.Int32) (Syntax: 'x')
+                Right: ILiteralExpression (Text: 10) (OperationKind.LiteralExpression, Type: System.Int32, Constant: 10) (Syntax: '10')
+        Body: IBranchStatement (BranchKind.Break) (OperationKind.BranchStatement) (Syntax: 'break;')
+    ISwitchCase (1 case clauses, 1 statements) (OperationKind.SwitchCase) (Syntax: 'default:/*< ... break;')
+        Clauses: ISingleValueCaseClause (Equality operator kind: BinaryOperationKind.None) (CaseKind.Default) (OperationKind.SingleValueCaseClause) (Syntax: 'default:')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.cs]: >ISingleValueCaseClause (Equality operator kind: BinaryOperationKind.None) (CaseKind.Default) (OperationKind.SingleValueCaseClause) (Syntax: 'default:') [](start = 17, length = 151)

Does this node have a label symbol?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.cs]: Please open an issue to expose the label symbol

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 2 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Sure, will change.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.cs (hunk 13 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.cs]: Filed https://github.com/dotnet/roslyn/issues/20366

**Comment by mavasani** [comment]
Unrelated test failures in [windows_release_unit64_prtest](
https://ci.dot.net/job/dotnet_roslyn/job/features_ioperation/job/windows_release_unit64_prtest/107/)

```
Microsoft.CodeAnalysis.Editor.UnitTests.LinkedFiles.LinkedFileDiffMergingEditorTests.TestCodeActionPreviewAndApply
Microsoft.CodeAnalysis.Editor.UnitTests.MetadataAsSource.MetadataAsSourceTests+VisualBasic.BracketedIdentifierSimplificationTest
```

@dotnet/roslyn-infrastructure Are these known?

**Comment by mavasani** [comment]
@dotnet-bot retest windows_release_unit64_prtest please

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs [hunk 14]
```diff
@@ -0,0 +1,679 @@
...
+                Diagnostic(ErrorCode.ERR_RbraceExpected, ":").WithLocation(9, 50),
+                // CS0168: The variable 'y' is declared but never used
+                //             /*<bind>*/case /*</bind>*/const int y:
+                Diagnostic(ErrorCode.WRN_UnreferencedVar, "y").WithArguments("y").WithLocation(9, 49)
+            };
+
+            VerifyOperationTreeAndDiagnosticsForTest<CaseSwitchLabelSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+    }
+}
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: Do we need to test redundant cases?
```
        switch (1)
        {
            case int x:
                break;
            case int y:
                break;
            case object z:
                break;
        }
```

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs (hunk 14 above)
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IPatternSwitchCase.cs]: Sure, let me add it in a separate PR I am working on currently (https://github.com/dotnet/roslyn/pull/20364)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 2 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: I will do this change in https://github.com/dotnet/roslyn/pull/20364, I forgot to add this commit to this PR.
